src/cli/commands/force/lightning/local/setup.ts

```diff
@@ -145,21 +112,15 @@ export class Setup extends SfdxCommand {
         return Promise.resolve();
     }
 
-    private async executeSetup(setup: BaseSetup): Promise<SetupTestResult> {
-        setup.skipBaseRequirements = this.skipBaseRequirements;
-        setup.skipAdditionalRequirements = this.skipAdditionalRequirements;
-        return setup.executeSetup();
-    }
-
-    private setup(): BaseSetup {
-        if (!this.setupSteps) {
-            this.setupSteps = CommandLineUtils.platformFlagIsAndroid(
+    public get commandChecks(): CommandChecks {
+        if (!this._commandChecks) {
+            this._commandChecks = CommandLineUtils.platformFlagIsAndroid(
                 this.flags.platform
             )
-                ? new AndroidEnvironmentSetup(this.logger, this.flags.apilevel)
-                : new IOSEnvironmentSetup(this.logger);
+                ? new AndroidEnvironmentChecks(this.logger, this.flags.apilevel)
+                : new IOSEnvironmentChecks(this.logger);
         }
 
-        return this.setupSteps;
+        return this._commandChecks;
```

Comment:
I'm having a hard time trying to understand how the refactoring in this PR is going to play out in other repos. Since we don't want our commands to extent from `Setup` command anymore, does it mean that every command is now going to duplicate something similar to this code snippet? And how can we determine if a command should run the environment checks first or should just skip them?

Reply:
We can't enforce coding pattern through inheritance. This was what made `Setup` a culprit in hoarding requirements in the firsts place. I can look into ways so other commands can be coded within the guard rails and have well-known pattern to follow to have their own set of requirements. However, if a rigid standard is going to be brittle or come up short I would rather offer flexibility and have the coder of the sfdx command to opt-in to having requirements run, either through the approach shown here, or, any approach that he or she can come up with. If you have suggestion for putting a system in place for it I like to take a stab at that too.

Reply:
Well at this point I just would like to understand how the current approach is going to work in different repos. Our commands may be executed individually from the command line, or in tandem (e.g from the uber command). In the case of the former, we need to have a way for a command to execute the setup checks first and then do its thing, whereas for the latter we want a command to skip the check. Right now I don't quite clearly see how we will accomplish this after the refactor. But then again once we start consuming it in the other repos, it might become more clear.

Reply:
I'm going to add some commentary that spans different code bodies in this PR, but has roots in this conversation too, so I'm dumping it all here. 🙂

1. But we'll start here. I'm not a fan of every Command having its own `commandChecks` property. The requirements processor (`CommandChecks`) can effectively be represented as a static processor. It takes requirements, it executes them, it reports results. This should be a service, not tied to a particular object that needs to use it.
2. As a corollary but also a point of allocation of responsibility, `Checks` should be owned by Commands, not a `CommandChecks` process. The latter should take them as an input.
3. I'd personally prefer to keep the naming of the base objects around the concept of requirements. `CommandChecks` could be `RequirementsProcessor` or similar. `Checks` could be `RequirementsCollection` or something like that. It's all related to requirements, and I feel like the tack to "checks" unnecessarily divorces that context.

Reply:
I like the idea of having a service... but I think taking that approach would turn out to be much more involved and require more work. Right now `Requirements.execute()` runs through an array of requirements **and** prints out the results. If we move to a service approach then I think we should do the right thing and have the service only be in charge of running the checks + returning their results, and have the caller do the printing. I don't think the service should be in charge of printing the results on the screen.... but then again the way we want things to work is that the checks should be running in parallel and we print the results of a check as soon as it is finished. Maybe we would need use Observables from RxJS or just have a dirty service that runs the tasks and prints the results even though it shouldn't be printing the results.... The refactoring task is quickly becoming something really big.

Reply:
I agree in principle with the concept of a properly-partitioned service, like you describe. As an intermediate step, I was hoping to simply take the code we have and effectively make it standalone.

This is basically what the `BaseSetup` class does today. I'm proposing that our first refactor simply separate that engine out from the requirements themselves.